File: JarvisClient/UserClient.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using JarvisClient.Extensions;
using Microsoft.AspNetCore.SignalR.Client;
using Shared.Messages;
using Shared.SignalR;
using static Shared.FunctionalConsole;
using static Shared.Messages.AgentCommand;

namespace JarvisClient;

/// <summary>
/// Represents a user client in the Jarvis project, handling communication and commands.
/// </summary>
/// <param name="hub">The SignalR HubConnection used for client-server communication.</param>
/// <param name="browser">The instance of ProjectBrowser for file and directory management.</param>
public class UserClient(HubConnection hub, ProjectBrowser browser) : IUserClient
{
    /// <summary>
    /// Receives and displays a message from the server.
    /// </summary>
    /// <param name="message">The message to be received and displayed.</param>
    /// <returns>A Task representing the asynchronous operation of message reception.</returns>
    public Task ReceiveMessage(string message)
    {
        Console.WriteLine(message);
        return Task.CompletedTask;
    }
    
    /// <summary>
    /// Receives and processes a command from the server, executing appropriate actions in the ProjectBrowser.
    /// </summary>
    /// <param name="correlationId">The correlation ID for the command.</param>
    /// <param name="command">The command to be processed.</param>
    /// <returns>A Task representing the asynchronous operation of command processing.</returns>
    public async Task ReceiveCommand(string correlationId, AgentCommand command)
    {
        var response = "Something went wrong";
        try
        {
            Console.WriteLine($"Receiving command: {command}");

            var result = command switch
            {
                ListProjectsCommand => browser.ListProjects().Apply(Serialize),
                GetProjectDetailsCommand(var projectName) => browser.GetProjectDetails(projectName).Select(Serialize),
                ListProjectDirectoryCommand(var projectName, var path) => browser.ListProjectDirectory(projectName, path).Select(Serialize),
                OpenFileCommand(var projectName, var path) => browser.OpenFile(projectName, path),
                WriteFileCommand(var projectName, var filePath, var content, var mode) => browser.WriteFile(projectName, filePath, content, mode),
                TextReplaceSectionCommand(var projectName, var filePath, var sectionIdentifiers, var replacementContent) =>
                    browser.ReplaceSection(projectName, filePath, sectionIdentifiers, replacementContent),
                TextReplaceCommand(var projectName, var filePath, var search, var content) => browser.Replace(projectName, filePath, search, content),
                TextInsertBeforeCommand(var projectName, var filePath, var search, var content) => browser.InsertBefore(projectName, filePath, search, content),
                TextInsertAfterCommand(var projectName, var filePath, var search, var content) => browser.InsertAfter(projectName, filePath, search, content),
                RunUnitTestsCommand(var projectName, var filePath) => browser.RunUnitTests(projectName, filePath),
                _ => error("Unknown command")
            };

            response = result.IfError(x =>
            {
                WriteLine(x.Message);
                return x.Message;
            });
        }
        catch (Exception e)
        {
            WriteLine(e.Message);
            response = e.Message;
        }
        finally
        {
            await hub.InvokeAsync<IJarvisHub>(x => x.SendClientResponse(correlationId, response));
        }
    }

    /// <summary>
    /// Serializes a given value to a JSON string using predefined JsonSerializer options.
    /// </summary>
    /// <param name="value">The value to be serialized.</param>
    /// <returns>A Result containing the serialized JSON string or an error message if serialization fails.</returns>
    private static Result<string> Serialize<T>(T value)
    {
        try
        {
            return ok(JsonSerializer.Serialize(value, Options()));
        }
        catch (Exception e)
        {
            return error(e);
        }
    }

    /// <summary>
    /// Serializes a given value to a JSON string using predefined JsonSerializer options.
    /// </summary>
    /// <param name="value">The value to be serialized.</param>
    /// <returns>A Result containing the serialized JSON string or an error message if serialization fails.</returns>
    private static AsyncEffect<string> Serialize2<T>(T value)
    {
        try
        {
            return okAsyncEffect(JsonSerializer.Serialize(value, Options()));
        }
        catch (Exception e)
        {
            return errorAsyncEffect<string>(e.Message);
        }
    }

    /// <summary>
    /// Configures JsonSerializerOptions for serialization processes in the UserClient.
    /// </summary>
    /// <returns>Configured JsonSerializerOptions with specific settings for the UserClient.</returns>
    private static JsonSerializerOptions Options()
    {
        var options = new JsonSerializerOptions(JsonSerializerDefaults.Web)
        {
            PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower
        };
        options.Converters.Add(new JsonStringEnumConverter());

        return options;
    }
}

File: Client/UserClient.fs
module Client.UserClient

open System.Text.Json
open System.Threading.Tasks
open Client.Effect
open Client.IO
open Client.SignalR
open Common
open Common.SignalR
open FsToolkit.ErrorHandling
open Microsoft.AspNetCore.SignalR.Client

let receiveMessage message = printfn $"%s{message}"

let serialize<'a> (value: 'a) =
    try
        JsonSerializer.Serialize value |> Ok
    with e ->
        e |> GenericError |> Error

let serialize'<'a> = Result.bind serialize<'a> >> ValueTask<_>

let receiveCommand (hub: HubConnection) rt (correlationId: string) (command: AgentCommand) =
    task {
        let! response =
            match command with
            | ListProjectsCommand -> ProjectBrowser.listProjects rt |> serialize'
            | OpenProjectCommand cmd -> rt |> ProjectBrowser.getProjectDetails cmd.ProjectName |> serialize'
            | ListProjectDirectoryCommand cmd ->
                rt
                |> ProjectBrowser.listProjectDirectory cmd.ProjectName cmd.FolderPath
                |> serialize'
            | ReadFileCommand cmd -> rt |> ProjectBrowser.openFile cmd.ProjectName cmd.FilePath |> serialize'
            | WriteFileCommand cmd ->
                rt
                |> ProjectBrowser.writeFile cmd.ProjectName cmd.FilePath cmd.Content cmd.FileWriteMode
                |> serialize'
            | TextReplaceSectionCommand cmd ->
                rt
                |> ProjectBrowser.replaceSection cmd.ProjectName cmd.FilePath cmd.SectionIdentifiers cmd.Content
                |> serialize'
            | TextReplaceCommand cmd ->
                ProjectBrowser.replaceText cmd.ProjectName cmd.FilePath cmd.Search cmd.Content rt
                |> serialize'
            | TextInsertBeforeCommand cmd ->
                rt
                |> ProjectBrowser.insertBefore cmd.ProjectName cmd.FilePath cmd.Search cmd.Content
                |> serialize'
            | TextInsertAfterCommand cmd ->
                rt
                |> ProjectBrowser.insertAfter cmd.ProjectName cmd.FilePath cmd.Search cmd.Content
                |> serialize'
            | LoadPageCommand cmd ->
                task {
                    let! result = Web.loadPage (Url cmd.Url) rt
                    return Result.bind serialize result
                }
                |> ValueTask<Result<_, _>>

        let respond message =
            hub.invokeAsync<IHubService> _.SendClientResponse(correlationId, message)

        let! _ =
            match response with
            | Ok result ->
                printfn "Command executed. Sending response."
                respond result
            | Error error ->
                let str = EffectError.toString error
                printfn "Command failed."
                printfn $"%s{str}"
                respond str

        ()
    }
    :> Task

type RuntimeConstraint<'a when 'a :> ProjectIO and 'a :> FileIO and 'a :> WebIO> = 'a

type Client<'a when RuntimeConstraint<'a>>(hub: HubConnection, rt: 'a) =
    interface IClientService with
        member this.ReceiveCommand(correlationId, command) =
            receiveCommand hub rt correlationId command

        member this.ReceiveMessage(message) = printfn $"%s{message}"
